edgecop-rspec.yml

```diff
@@ -0,0 +1,11 @@
+RSpec/NestedGroups:
+  Enabled: false
+
+RSpec/MultipleMemoizedHelpers:
+  Enabled: false
+
+RSpec/MultipleExpectations:
+  Enabled: false
+
+RSpec/LetSetup:
+  Enabled: false
```

Comment:
I strongly disagree.

**_stop us from using `let!` when we have a useless assignment to variable (i.e. when that let isn't being used anywhere in the spec)_**
The implication that because the `let!` isn't called in the test explicitly it's therefore useless is fallacy. It has plenty of uses, principally setting up states. The solution you're outline is a direct work around to this and can be equally abused.
This is by no means a settled issue and if it were they why don't rspec deprecate `let!` all together.
 See: https://github.com/rubocop/rubocop-rspec/issues/94
IMO, all this does is turn one line of code into multiple lines. Can this be abused? Yes, but so can the alternative. It also removes the ability to override the variable in a later context. You can overwrite `let!` but not `before`.

At the end of the day, this is preference. Keeping this enabled is effectively enforcing that preference on everyone. This is still not at present the defacto way everyone writes code. Until it's mandated that this is enabled, the default has to be keep this disabled

Reply:
But we don't want to override variables I thought? That goes against structuring your spec in such a way that once something has been defined you can have higher confidence in it being consistent then for any nested contexts.

This isn't saying that you can never use `let!` (although I haven't ever come across a need for it), it's just saying that if you're not referencing it in your test set up then you definitely don't need to use let! and you should instead put the setup you were doing in `let!` in an appropriately placed before block. Avoiding overriding in this case is actually a further bonus of taking this approach and having spec structure more closely follow the logical flow of the code under test.

Reply:
Whats wrong with overriding exactly? :p

Reply:
Oh no! I didn't realise how controversial this would be! I'm personally not knowledgeable enough about Rspec to have a good opinion here, but I'm reaching out to a few more backendy people to see how they feel

Reply:
I agree with Peter on this.

This was discussed before in [The Bikeshed channel](https://teams.microsoft.com/l/message/19:320e455af1f6458e81a040bfc6fe2df1@thread.skype/1665760171422?tenantId=b503d5f8-66cb-4c4d-8a56-053fbd4b0ebe&groupId=32a06cda-ba10-4ff5-939a-0077651b9014&parentMessageId=1665760171422&teamName=dev&channelName=The%20Bikeshed&createdTime=1665760171422&allowXTenantAccess=false) and the people who contributed to that thread, were in favour of using `let!`. 

@ernest4 Do you still agree with using `let!` or have you changed your mind since you contributed to that thread?

Reply:
Ok for the sake of getting these changes out then let's ignore it for now. I'll gather some examples of where I see things being misused or making life difficult as I come across them again and come back with a more convincing argument. :P I won't be using `let!` myself and definitely won't be encouraging anyone else to use it or suggesting refactors to use `let!` though.

Reply:
As long time **let!** user, after first seeing Conor be against **let!** months ago, I've been writing code without **let!** to see how it goes and honestly i dont see the problem of banning **let!**. I haven't missed it one bit.

there's nothing **before** block cant replicate that **let!** does.

_"IMO, all this does is turn one line of code into multiple lines"_

and that's the main difference between **let!** and **before**, in _favour_ of **before**, not against imo. both of them can be abused, sure, but one takes more effort to do that and thus acts as higher barrier to overcome. To turn **let** to **let!** is ludicrously easy and visually you don't event notice, it can be done on subconscious level almost - easy to abuse and hard to spot the abuse. With **before**, you're making a conscious decision that takes more for to execute and easier to spot. 

So given that **let!** and **before** have same functionality, im in favour of **before** because of extra lines to help deter abuse.

 _"You can overwrite let! but not before."_
 
 How's that? You can define **before** in nested contexts that does things different to **before** in higher contexts, just the same as you do with **let!**. Any examples of that not working? As mentioned, I haven't run into any issues avoiding **let!**.
 
_"At the end of the day, this is preference."_

I weakly disagree. _Disagree_ because as mentioned **before** has the advantage of disincentivising abuse by forcing the writer to consciously commit to writing more lines to get the behaviour. _Weakly_ because the benefit carries to first instance of **before**, after that copy paste makes it easy again...but at least, still stands out more in PRs.

Reply:
Maybe a bit late, but I'm also on team `let/before`.
Agree with everything @ernest4 said above.

Reply:
Thanks for everyone's input! As an initial step, let's keep things as they are and not ban `let!` for now. We're already making a pretty drastic change by enforcing all new PRs to have 0 Rubocop violations from now on. We can revisit `let!` vs `let/before` in a while, once the dust from enforcing Rubocop has settled.

Reply:
> you're not overriding a before block with another before in a higher context, you're just executing both so this is just worse

how is it worse? it's the exact same. if you define two **let!** with one overriding the other, both will be executed as well - that's what the **!** guarantees after all.

**let!** and **let/before** are literally, functionally, identical afaik

But i agree with your overall point that it may not have _enough_ utility to be enforced. My only point was that in absolute terms, **let/before** has more utility and is better construct to use imo. but as you said, does it have _enough_ utility to justify heavy handed approach? ehhh.... im not convinced on that either :D